### PR TITLE
Role generation with description

### PR DESCRIPTION
**Pull request by Appointat**
## Description

Describe your changes in detail.

## Motivation and Context

- [x] I have raised an issue to propose this change ([Feature Request](https://github.com/camel-ai/camel/issues/239 for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [x] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [x] I have updated the documentation accordingly.

## Output of the new example
``` shell
AI Assistant sys message:
BaseMessage(role_name='Software Engineer', role_type=<RoleType.ASSISTANT: 'assistant'>, meta_dict={'task': 'Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.', 'assistant_role': 'Software Engineer', 'user_role': 'Financial Analyst', 'assistant_description': "Designing, developing, and testing software for the trading bot; creating algorithms and data structures to optimize the bot's performance; troubleshooting and debugging the bot; monitoring the bot's performance and making adjustments as needed.", 'user_description': "Analyzing financial data and market trends; researching and forecasting stock prices; developing strategies for the trading bot; monitoring the bot's performance and making adjustments as needed."}, content="===== ROLES WITH DESCRIPTION =====\nFinancial Analyst and Software Engineer are collaborating to complete a task: Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.\nFinancial Analyst's competencies, professional characteristics, duties and workflows to complete the task: Analyzing financial data and market trends; researching and forecasting stock prices; developing strategies for the trading bot; monitoring the bot's performance and making adjustments as needed.\nSoftware Engineer's competencies, professional characteristics, duties and workflows to complete the task: Designing, developing, and testing software for the trading bot; creating algorithms and data structures to optimize the bot's performance; troubleshooting and debugging the bot; monitoring the bot's performance and making adjustments as needed.\n===== RULES OF ASSISTANT =====\nNever forget you are a Software Engineer and I am a Financial Analyst. Never flip roles! Never instruct me!\nWe share a common interest in collaborating to successfully complete a task.\nYou must help me to complete the task.\nHere is the task: Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.. Never forget our task!\nI must instruct you based on your expertise and my needs to complete the task.\n\nI must give you one instruction at a time.\nYou must write a specific solution that appropriately solves the requested instruction and explain your solutions.\nYou must decline my instruction honestly if you cannot perform the instruction due to physical, moral, legal reasons or your capability and explain the reasons.\nUnless I say the task is completed, you should always start with:\n\nSolution: <YOUR_SOLUTION>\n\n<YOUR_SOLUTION> should be very specific, include detailed explanations and provide preferable detailed implementations and examples and lists for task-solving.\nAlways end <YOUR_SOLUTION> with: Next request.")

AI User sys message:
BaseMessage(role_name='Financial Analyst', role_type=<RoleType.USER: 'user'>, meta_dict={'task': 'Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.', 'assistant_role': 'Software Engineer', 'user_role': 'Financial Analyst', 'assistant_description': "Designing, developing, and testing software for the trading bot; creating algorithms and data structures to optimize the bot's performance; troubleshooting and debugging the bot; monitoring the bot's performance and making adjustments as needed.", 'user_description': "Analyzing financial data and market trends; researching and forecasting stock prices; developing strategies for the trading bot; monitoring the bot's performance and making adjustments as needed."}, content='===== ROLES WITH DESCRIPTION =====\nFinancial Analyst and Software Engineer are collaborating to complete a task: Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.\nFinancial Analyst\'s competencies, professional characteristics, duties and workflows to complete the task: Analyzing financial data and market trends; researching and forecasting stock prices; developing strategies for the trading bot; monitoring the bot\'s performance and making adjustments as needed.\nSoftware Engineer\'s competencies, professional characteristics, duties and workflows to complete the task: Designing, developing, and testing software for the trading bot; creating algorithms and data structures to optimize the bot\'s performance; troubleshooting and debugging the bot; monitoring the bot\'s performance and making adjustments as needed.\n===== RULES OF USER =====\nNever forget you are a Financial Analyst and I am a Software Engineer. Never flip roles! You will always instruct me.\nWe share a common interest in collaborating to successfully complete a task.\nI must help you to complete the task.\nHere is the task: Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.. Never forget our task!\nYou must instruct me based on my expertise and your needs to solve the task ONLY in the following two ways:\n\n1. Instruct with a necessary input:\nInstruction: <YOUR_INSTRUCTION>\nInput: <YOUR_INPUT>\n\n2. Instruct without any input:\nInstruction: <YOUR_INSTRUCTION>\nInput: None\n\nThe "Instruction" describes a task or question. The paired "Input" provides further context or information for the requested "Instruction".\n\nYou must give me one instruction at a time.\nI must write a response that appropriately solves the requested instruction.\nI must decline your instruction honestly if I cannot perform the instruction due to physical, moral, legal reasons or my capability and explain the reasons.\nYou should instruct me not ask me questions.\nNow you must start to instruct me using the two ways described above.\nDo not add anything else other than your instruction and the optional corresponding input!\nKeep giving me instructions and necessary inputs until you think the task is completed.\nWhen the task is completed, you must only reply with a single word <CAMEL_TASK_DONE>.\nNever say <CAMEL_TASK_DONE> unless my responses have solved your task.')

Original task prompt:
Develop a trading bot for the stock market.

Specified task prompt:
Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.

Final task prompt:
Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.

AI User: Financial Analyst

Instruction: Develop a sentiment analysis algorithm to analyze social media trends in real-time.
Input: None


AI Assistant:Software Engineer

Solution: To develop a sentiment analysis algorithm to analyze social media trends in real-time, we can follow the following steps:

1. Data Collection: Gather a large dataset of social media posts or tweets related to the financial market. This dataset should include both positive and negative sentiment examples.

2. Preprocessing: Clean the collected data by removing any irrelevant information such as URLs, hashtags, and special characters. Also, perform tokenization to split the text into individual words or tokens.

3. Feature Extraction: Convert the preprocessed text into numerical features that can be used by machine learning algorithms. Common techniques for feature extraction include bag-of-words, TF-IDF, and word embeddings like Word2Vec or GloVe.

4. Labeling: Manually label the collected data with sentiment labels (positive or negative). This step is necessary to train a supervised machine learning model.
(not finished...)
```

### CODE REVIEW

**Comment by lightaime** [comment]
Thanks for the PR! @Appointat

File: camel/agents/chat_agent.py
Comment (Obs01ete) [inline camel/agents/chat_agent.py:355]: Why do we need `step_completion` method if we already have `step`?
THe contents seem to be a simplified duplicate of step. How can we make it DRY?

File: camel/agents/task_agent.py
Comment (Obs01ete) [inline camel/agents/task_agent.py:381]: Move `RoleAssignmentAgent` to a separate py file.

File: camel/agents/task_agent.py
Comment (Obs01ete) [inline camel/agents/task_agent.py:448]: varialbles must all be snake case

File: camel/models/model_factory.py
Comment (Obs01ete) [inline camel/models/model_factory.py:49]: Why do we need to add a deprecated model?
I believe `ModelType.GPT_3_5_TURBO` can do simple completion as well (not only chat completion).

File: camel/models/openai_model.py
Comment (Obs01ete) [inline camel/models/openai_model.py:52]: Raise on empty messages and keys not there.

File: camel/societies/role_playing.py
Comment (Obs01ete) [inline camel/societies/role_playing.py:124]: Extract into `assistant_description` and `user_description` variable

File: camel/societies/role_playing.py
Comment (Obs01ete) [inline camel/societies/role_playing.py:231]: @lightaime I do not like how the constructor grows to 20+ parameters. This break the cohesion. RolePlaying now is responsible for knowing about all the agents and their details. We do not want to create a god class here.

File: camel/typing.py
Comment (Obs01ete) [inline camel/typing.py:33]: Let's not add the deprecated model. Use whatever model is already there, they all support completion.

File: examples/ai_society/role_playing_with_role_generation.py
Comment (Obs01ete) [inline examples/ai_society/role_playing_with_role_generation.py:26]: Add a test to this eample.

File: camel/agents/task_agent.py
Comment (Obs01ete) [inline camel/agents/task_agent.py:381]: Add a test for this newly introduced class.

**Review by Obs01ete — CHANGES_REQUESTED** [review]
Looks like a good feature. For now you need to address the comments. Especially about how to avoid adding step_completion method.

File: camel/typing.py
Comment (Obs01ete) [inline camel/typing.py:22]: How can a role type be of value "role description". I do not get the concept.

File: camel/societies/role_playing.py
Comment (Obs01ete) [inline camel/societies/role_playing.py]: Do not use empty strings. It was good before, with Optional. Change to None and defaults to None as well.

File: camel/agents/chat_agent.py
Comment (Obs01ete) [inline camel/agents/chat_agent.py:385]: Asserts are not allowed in the package code.

**Review by Obs01ete — CHANGES_REQUESTED** [review]

File: camel/agents/chat_agent.py
Comment (Appointat) [reply camel/agents/chat_agent.py:355]: # used for output structure string.
I would like to fill in the blank in the prompt using `OpenAI Completion API` [http:](https://platform.openai.com/docs/api-reference/completions). For example:
## prompt
```
Domain expert 1: <|blank|>
Associated competencies, professional characteristics, duties and workflows: <|blank|>. End
```
And then the completion API will return the response, for example:
```
Domain expert 1: Data Scientist
Associated competencies, professional characteristics, duties and workflows: Mastery in statistical analysis. Proficiency in programming languages such as Python and R. End
```
What is more, you will see the API of `Completion` and `ChatCompletion` is different, so in my personal opinion, it is not a duplication.

File: camel/agents/task_agent.py
Comment (Appointat) [reply camel/agents/task_agent.py]: Thanks, done. : )

File: camel/models/model_factory.py
Comment (Appointat) [reply camel/models/model_factory.py:49]: As I have chosen to use the `Completion` API, the model is related to `TEXT_DAVINCI_003`, and its quality of response is much better than `TURBO` in the case of filling in the blanks. And `Completion` is not supported by the `TURBO` model.

File: camel/models/openai_model.py
Comment (Appointat) [reply camel/models/openai_model.py:59]: Thanks, done : )

File: camel/societies/role_playing.py
Comment (Appointat) [reply camel/societies/role_playing.py:246]: Emmm, it would be an issue. At least, I think `sth_description` is necessary, as it is available to define the actions/personalities/rules of the agents in the future. Need to be discussed.

File: camel/typing.py
Comment (Appointat) [reply camel/typing.py:34]: As you mentioned it, I will try to move the `TEXT_DAVINCI_003`.

File: examples/ai_society/role_playing_with_role_generation.py
Comment (Appointat) [reply examples/ai_society/role_playing_with_role_generation.py:26]: Thanks, I have added a `pytest` test in `examples\test\test_ai_society_with_role_generation.py`. But I am not sure if I do right. Your comments will be appreciated, thanks.

File: camel/agents/chat_agent.py
Comment (Appointat) [reply camel/agents/chat_agent.py]: Thanks, done : )

File: camel/agents/task_agent.py
Comment (Appointat) [reply camel/agents/task_agent.py]: Thanks, I created `examples\test\test_ai_society_with_role_generation.py`, but I am not so sure if I do right.

File: camel/societies/role_playing.py
Comment (Appointat) [reply camel/societies/role_playing.py:139]: Thanks, done.

File: camel/typing.py
Comment (Appointat) [reply camel/typing.py:22]: Considering your comment, I added a new type `DescriptionType`. in this type, it is available to define kinds of description types (like, skills, competences, duties). At present, I only defined a default description type:
```
# camel\typing.py
class DescriptionType(Enum):
    DEFAULT = "default_description"
```

File: camel/typing.py
Comment (Appointat) [reply camel/typing.py]: Now I have removed this model.

File: camel/models/model_factory.py
Comment (Appointat) [reply camel/models/model_factory.py]: Now I have removed the model according to your comment.

File: camel/agents/chat_agent.py
Comment (Appointat) [reply camel/agents/chat_agent.py]: Now I have removed the model (using `TURBO` rather than `DAVINCI`) according to your comment.

File: camel/typing.py
Comment (Obs01ete) [inline camel/typing.py:67]: What is the purpose of DescriptionType? WHy do we need just DEFAULD? THey use of it is not clear. You can put it as a docstring of DescriptionType which is needed anyway.

File: examples/ai_society/role_playing_with_role_generation.py
Comment (Obs01ete) [reply examples/ai_society/role_playing_with_role_generation.py:26]: @Appointat Check out the test for an exaple here: `examples\test\test_code_example.py`. It is as simple as calling main()

File: examples/test/test_ai_society_with_role_generation.py
Comment (Obs01ete) [inline examples/test/test_ai_society_with_role_generation.py:23]: Rework as recommended above.

**Review by Obs01ete — CHANGES_REQUESTED** [review]
This is way better than the previous version.

File: examples/test/test_ai_society_with_role_generation.py
Comment (Appointat) [reply examples/test/test_ai_society_with_role_generation.py:23]: Thanks, done : )

File: camel/agents/task_agent.py
Comment (Appointat) [reply camel/agents/task_agent.py]: Done

File: camel/typing.py
Comment (Obs01ete) [reply camel/typing.py:22]: @lightaime can you take a look? We need someone smart to understand what is going on here.

**Review by Obs01ete — CHANGES_REQUESTED** [review]
All good apart from DescriptionType the concept of which I do not understand. I will approve once @lightaime approves

File: camel/typing.py
Comment (Appointat) [reply camel/typing.py:22]: Please wait, I will modify the structure with this type tonight. And then, I will explain the changes I have made here. Thanks for your patience.

File: examples/ai_society/role_playing_with_role_generation.py
Comment (Appointat) [reply examples/ai_society/role_playing_with_role_generation.py:25]: Thanks, done : )

File: camel/typing.py
Comment (Appointat) [reply camel/typing.py:67]: The `DescriptionType` class is designed to categorize and differentiate various types or requirements of descriptions. For instance, there might be situations where we need descriptions specific to skills, duty, capability, or even personality traits. This is the rationale behind defining a separate class for description types.

``` python
class DescriptionType(Enum):
    DEFAULT = "default_description"
```

In the provided `DescriptionType` class, we've only defined a `DEFAULT` description template. The template serves as a basic structure to describe the collaboration between different roles:

```
===== ROLES WITH DESCRIPTION =====
{user_role} and {assistant_role} are collaborating to complete a task: {task}
{user_role}'s competencies, professional characteristics, duties and workflows to complete the task: {user_description}
{assistant_role}'s competencies, professional characteristics, duties and workflows to complete the task: {assistant_description}
```

This template illustrates a general collaboration scenario between the user and an assistant. However, by having a `DescriptionType` class, we lay the foundation for adding more specific description types in the future. For example, we can introduce types like `SKILL`, `DUTY`, `CAPABILITY`, and `PERSONALITY` to **offer more detailed and varied descriptions** based on different contexts or requirements.

File: camel/societies/role_playing.py
Comment (Appointat) [reply camel/societies/role_playing.py:247]: I have explained in this [link](https://github.com/camel-ai/camel/pull/240#discussion_r1290562032), please review it thanks.

File: camel/typing.py
Comment (Obs01ete) [reply camel/typing.py:67]: @Appointat Thanks for the detailed description of the idea behind the DescriptionType. I'd prefer to not introduce it in this MR, and to keep it in the backlog until the moment we have more than one DescriptionType. It may happen that we'll never need DescriptionType and by postponing this change we will keep the code simpler and easier to understand and use. Once you come up with a PR with 2+ options for DescriptionType we will consider merging it.

Comment (Appointat) [reply camel/typing.py:67]: Thanks for your response, I will modify these changes this weekend. I am working on another task, so, sorry for not replying in time.

**Comment by lightaime** [comment]
@Appointat Thanks for the PR. Some tests fail because of GitHub secrets. This is a drawback of using `fork-and-pull-request` workflow. An alternative way to create PR is to use `checkout-and-pull-request` workflow. I just added an instruction [here](https://github.com/camel-ai/camel/blob/53b7109c71144616d4aaae30504b21e8602bab56/CONTRIBUTING.md) for your reference.

File: camel/typing.py
Comment (Appointat) [reply camel/typing.py]: Resolved, I removed the `DescriptionType`. Thanks

**Pull request closed without merging**
